lib/ABI/FunctionType.cpp

```diff
@@ -110,6 +95,35 @@ static bool verifyReturnValueLocation(const abi::Definition &D) {
   return true;
 }
 
+/// Replace all the references to `OldKey` with the references to
+/// the newly added `NewType`. It also erases the old type.
+///
+/// \param OldKey The type references of which should be replaced.
+/// \param NewType The new type to replace the references to. It is added to
+///        the model implicitly, no need to do so before calling this function.
+/// \param Model The tuple tree where replacement should take place in.
+///
+/// \return The new path to the added type.
+static model::TypePath replaceAllUsesWith(const model::Type::Key &OldKey,
+                                          const model::TypePath &NewTypePath,
+                                          TupleTree<model::Binary> &Model) {
+  auto Visitor = [&NewTypePath, &OldKey](model::TypePath &Visited) {
+    if (!Visited.isValid())
+      return; // Ignore empty references
+
+    const model::Type *Current = Visited.getConst();
+    revng_assert(Current != nullptr);
+
+    if (Current->key() == OldKey)
+      Visited = NewTypePath;
+  };
+  Model.visitReferences(Visitor);
+
+  Model->Types.erase(OldKey);
+
+  return NewTypePath;
+}
```

Comment:
We have `Binary::replaceReferences`. See if it fits your purpose or if you can extend it to do so.

Reply:
Sadly it would have to be something like
```cpp
// STLExtras.h

  template<typename Fn, typename R, typename ...Args>
  concept invocable_r = std::is_invocable_r<R, Fn, Args...>;
  
// TupleTree.h

  template<StrictSpecializationOf<TupleTreeReference> TTR,
           invocable_r<bool, const TTR &> Condition>
  void replaceReferencesIf(TTR &NewReference, Condition &&Lambda) {
    auto Visitor = [&Lambda, &NewReference](const TTR &Reference) {
      if (Lambda(Reference))
        Reference = NewReference;
    };
    visitReferences(Visitor);
    evictCachedReferences();
  }
  
// FunctionType.cpp

  TupleTree<model::Binary> MyBinary = getItFromSomewhere();
  auto CheckTheKey = [&OldKey](const model::TypePath &Reference) -> bool {
    return OldKey == Reference.getConst()->key();
  };
  MyBinary.replaceReferencesIf(NewTypePath, CheckTheKey);

```

Which is honestly dislike quite a bit. Do you think it's still worth it? Or should I just add a `evictCachedReferences()` invocation to my current version and call it a day?

Reply:
Why do you dislike it? It's similar to `erase_if`.

Reply:
> Why do you dislike it?

Mostly because I'm starting to feel like that's too many distinct lambda parameters per function call (don't mind me, a past trauma).

But seriously, when there's a lambda capturing another lambda, maybe we're a step too far, don't you think?

> It's similar to `erase_if`.

Yep, I intentionally wrote it this way.

> I'll leave this to you.

Which way do you prefer? The multiple-lambda one?

Reply:
You're viewing it in your specific use case here, but from my POV I'd be happy to have `replaceReferencesIf`, in general.

Reply:
If you can think of other places that can benefit from it then sure, I don't mind moving it up, lambdas or no lambdas.

Reply:
This is done: https://github.com/revng/revng/pull/232/commits/65cc91e7b7c49e399ccbb56d52d330d129d2cdb2 and https://github.com/revng/revng/pull/232/commits/2696027c0cdd38fcd67568f9cafa3e59c3841d7f.